devices/usbc/ctrl.c

```diff
@@ -0,0 +1,400 @@
+};
+
+
+static int ctrl_allocBuff(int endpt, int dir)
+{
+	ctrl_common.data->endpts[endpt].buf[dir].buffer = usbclient_allocBuff(USB_BUFFER_SIZE);
+
+	if (ctrl_common.data->endpts[endpt].buf[dir].buffer == NULL)
+		return ERR_MEM;
+
+	ctrl_common.data->endpts[endpt].buf[dir].len = 0;
+
+	return ERR_NONE;
+}
+
+
+static void *ctrl_allocQtdMem(void)
+{
+	int i;
+
+	if (!ctrl_common.qtdOffs) {
+		ctrl_common.dc->dtdMem = usbclient_allocBuff(USB_BUFFER_SIZE);
+
+		if (ctrl_common.dc->dtdMem == NULL)
+			return NULL;
+
+		for (i = 0; i < USB_BUFFER_SIZE; ++i)
+			((char *)ctrl_common.dc->dtdMem)[i] = 0;
+	}
+
+	return  (void *)(ctrl_common.dc->dtdMem + 0x40 * (ctrl_common.qtdOffs++));
+}
+
+
+static int ctrl_dtdInit(int endpt, int inQH, int outQh)
+{
+	dtd_t *dtd;
+	int qh = endpt * 2;
+
+	if (!endpt)
+		return -1;
+
+	/* initialize dtd for selected endpoints */
+	if (outQh) {
+		dtd = ctrl_allocQtdMem();
+		if (dtd == NULL)
+			return ERR_MEM;
+
+		ctrl_common.dc->endptqh[qh].base = (((u32)dtd) & ~0xfff);
+		ctrl_common.dc->endptqh[qh].size = 0x40;
+		ctrl_common.dc->endptqh[qh].head = dtd;
+		ctrl_common.dc->endptqh[qh].tail = dtd;
+	}
+
+	if (inQH) {
+		dtd = ctrl_allocQtdMem();
+		if (dtd == NULL)
+			return ERR_MEM;
+
+		ctrl_common.dc->endptqh[++qh].base = (((u32)dtd) & ~0xfff) + (64 * sizeof(dtd_t));
+		ctrl_common.dc->endptqh[qh].size = 0x40;
+		ctrl_common.dc->endptqh[qh].head = dtd + 64;
+		ctrl_common.dc->endptqh[qh].tail = dtd + 64;
+	}
+
+	return ERR_NONE;
+}
+
+
+static int ctrl_initEndptQh(int endpt, int dir, endpt_data_t *endpt_init)
+{
+	u32 setup = 0;
+	int qh = endpt * 2 + dir;
+
+	ctrl_allocBuff(endpt, dir);
+
+	ctrl_common.dc->endptqh[qh].caps =  endpt_init->caps[dir].max_pkt_len << 16;
+	ctrl_common.dc->endptqh[qh].caps |= endpt_init->caps[dir].ios << 15;
+	ctrl_common.dc->endptqh[qh].caps |= endpt_init->caps[dir].zlt << 29;
+	ctrl_common.dc->endptqh[qh].caps |= endpt_init->caps[dir].mult << 30;
+	ctrl_common.dc->endptqh[qh].dtd_next = 1;
+
+	setup |= endpt_init->ctrl[dir].data_toggle << (6 + dir * 16);
+	setup |= endpt_init->ctrl[dir].type << (2 + dir * 16);
+
+	return setup;
+}
+
+
+int ctrl_endptInit(int endpt, endpt_data_t *endpt_init)
+{
+	s8 i;
+	int res = ERR_NONE;
+	u32 setup = 0;
+
+	if (endpt == 0)
+		return -1;
+
+	if ((res = ctrl_dtdInit(endpt, endpt_init->caps[USB_ENDPT_DIR_IN].init, endpt_init->caps[USB_ENDPT_DIR_OUT].init)) != ERR_NONE)
+		return res;
+
+	for(i = 0; i < ENDPOINTS_DIR_NB; ++i) {
+		if (endpt_init->caps[i].init)
+			setup |= ctrl_initEndptQh(endpt, i, endpt_init);
+	}
+
+	*(ctrl_common.dc->base + endptctrl0 + endpt) = setup;
+
+	for(i = 0; i < ENDPOINTS_DIR_NB; ++i) {
+		if (endpt_init->caps[i].init)
+			*(ctrl_common.dc->base + endptctrl0 + endpt) |= 1 << (7 + i * 16);
+	}
+
+	return res;
+}
+
+
+int ctrl_endpt0Init(void)
+{
+	int i;
+	u32 qh_addr;
+
+	ctrl_allocBuff(0, USB_ENDPT_DIR_IN);
+
+	ctrl_allocBuff(0, USB_ENDPT_DIR_OUT);
```

Comment:
In theory we might get `ERR_MEM` from them

Reply:
Thanks, you have right.
Fixed.